dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,199 @@ public Pattern getParameterPattern() {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
+        }
+    }
+
+    private class DatabaseProcedure extends DatabaseExecutable {
+
+        private DatabaseProcedure(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getProcedures(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseExecutable {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseObjectName {
+        String schemaName, objectName;
+
+        private DatabaseObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    DatabaseObjectName createDbObjectNameWithCurrentDefaultSchema(String objName) throws SQLException {
+        return new DatabaseObjectName(getConnection().getSchema(), objName);
```

Comment:
Yes, some kind of factory/builder/parser class or method comes to my mind too. Parsing methods or c-tors inside the same class itself can't solve such problem. Or it could be a static method with a different name.

It's something to look into when we extend the usage to another place.